include/aikido/control/Executor.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef AIKIDO_CONTROL_EXECUTOR_HPP_
+#define AIKIDO_CONTROL_EXECUTOR_HPP_
+
+#include <chrono>
+#include <future>
+#include <set>
+#include <vector>
+
+#include "aikido/common/ExecutorThread.hpp"
+#include "aikido/common/pointers.hpp"
+
+namespace aikido {
+namespace control {
+
+AIKIDO_DECLARE_POINTERS(Executor)
+
+enum class ExecutorType
+{
+  STATE = -1,
+  POSITION = 0,
+  VELOCITY = 1,
+  EFFORT = 2,
+  TRAJECTORY = 3,
+  OTHER = 4
+};
+
+// Parameter defaults
+constexpr std::chrono::milliseconds defaultThreadRate{10};
+
+/// Abstract class for executing commands
+class Executor
+{
+public:
+  virtual ~Executor()
+  {
+    stop();
+  }
+
+  /// Constructor.
+  /// \param[in] types Vector of controller types
+  /// \param[in] joints Vector of joint names this Executor acts upon
+  /// \param[in] threadRate (Optional) How often to call step()
+  Executor(
+      const std::vector<ExecutorType> types,
+      const std::vector<std::string> joints,
+      std::chrono::milliseconds threadRate = defaultThreadRate)
+    : mThreadRate(threadRate), mThread(nullptr), mTypes(types), mJoints(joints)
+  {
+  }
+
+  Executor(
+      const ExecutorType type,
+      const std::vector<std::string> joints,
+      std::chrono::milliseconds threadRate = defaultThreadRate)
+    : Executor(std::vector<ExecutorType>{type}, joints, threadRate)
+  {
+  }
+
+  /// Get primary ExecutorType
+  ExecutorType getType()
+  {
+    return mTypes[0];
+  }
+
+  /// Get all of this Executor's ExecutorTypes
+  std::vector<ExecutorType> getTypes()
+  {
+    return mTypes;
+  }
+
+  /// Get list of joints needed by this Executor
+  std::vector<std::string> getJoints()
+  {
+    return mJoints;
+  }
+
+  /// Step to a point in time.
+  /// \note \c timepoint can be a time in the future to enable faster than
+  /// real-time execution.
+  ///
+  /// \param timepoint Time to simulate to
+  virtual void step(const std::chrono::system_clock::time_point& timepoint) = 0;
+
+  /// Start the underlying ExecutorThread
+  void start()
+  {
+    if (!mThread)
+    {
+      mThread = std::make_unique<aikido::common::ExecutorThread>(
+          std::bind(&Executor::spin, this), mThreadRate);
+    }
+  }
+
+  /// Stops the underlying ExecutorThread
+  void stop()
+  {
+    if (mThread)
+    {
+      mThread->stop();
+      mThread.reset();
+    }
+  }
+
+private:
+  /// Call to spin first to pass current time to step
```

Comment:
Actually, the reason this function is necessary is that, in order to enable real-time execution using an Executor's ExecutorThread, it needs a way to pass the current time into step(). This can't be done with std::bind (which would only pass the time point of construction) or a default argument (which also only passes the time point of construction). It also acts as a check to see if a joint has been requested of the thread, and if so, to not bother executing step().